lib/image-stream/parse-partitions.js

```diff
@@ -0,0 +1,174 @@
+const MAX_BLOCK_SIZE = 4096;
+
+/**
+ * @summary Attempt to parse the GPT from various block sizes
+ * @private
+ * @param {Buffer} buffer - Buffer
+ * @returns {GPT|null}
+ */
+const parseGPT = (buffer) => {
+
+  let blockSize = INITIAL_BLOCK_SIZE;
+  let gpt = null;
+
+  // Attempt to parse the GPT from several offsets,
+  // as the block size of the image may vary (512,1024,2048,4096);
+  // For example, ISOs will usually have a block size of 4096,
+  // but raw images a block size of 512 bytes
+  while (blockSize <= MAX_BLOCK_SIZE) {
+    try {
+      gpt = GPT.parse(buffer.slice(blockSize));
+    } catch (error) {
+      // Ignore error
+    }
+    if (gpt) {
+      return gpt;
+    }
+    blockSize += blockSize;
+  }
+
+  return gpt;
+
+};
+
+/**
+ * @summary Attempt to parse the MBR & GPT from a given buffer
+ * @private
+ * @param {Object} image - Image metadata
+ * @param {Buffer} buffer - Buffer
+ * @returns {undefined}
+ */
+const parsePartitions = (image, buffer) => {
+
+  let mbr = null;
+  let gpt = null;
+
+  try {
+    mbr = MBR.parse(buffer);
+    image.hasMBR = true;
+    if (mbr.getEFIPart()) {
+      image.hasGPT = true;
+      gpt = parseGPT(buffer);
+    }
+  } catch (error) {
+    // Ignore error
+  }
+
+  // As MBR and GPT partition entries have a different structure,
+  // we normalize them here to make them easier to deal with and
+  // avoid clutter in what's sent to analytics
+  if (image.hasGPT) {
+    image.partitions = _.map(gpt.partitions, (partition) => {
+      return {
+        type: partition.type.toString(),
+        id: partition.guid.toString(),
+        name: partition.name,
+        firstLBA: partition.firstLBA,
+        lastLBA: partition.lastLBA,
+        extended: false
+      };
+    });
+  } else if (image.hasMBR) {
+    image.partitions = _.map(mbr.partitions, (partition) => {
+      return {
+        type: partition.type,
+        id: null,
+        name: null,
+        firstLBA: partition.firstLBA,
+        lastLBA: partition.lastLBA,
+        extended: partition.extended
+      };
+    });
+  }
+
+};
+
+/**
+ * @summary Attempt to read the MBR and GPT from an imagestream
+ * @description
+ *   This operation will consume the first `MAX_BYTES`
+ *   of the stream and then destroy the stream.
+ * @param {Object} image - image metadata
+ * @returns {Promise}
+ */
+module.exports = (image) => {
+  return new Bluebird((resolve, reject) => {
+
+    const chunks = [];
+    let length = INITIAL_LENGTH;
+    let destroyed = false;
+
+    image.hasMBR = false;
+    image.hasGPT = false;
+
+    let stream = image.stream.pipe(image.transform);
+
+    stream.on('error', reject);
+
+    // We need to use the "old" flowing mode here,
+    // as some dependencies don't implement the "readable"
+    // mode properly (i.e. bzip2)
+    stream.on('data', (chunk) => {
+      chunks.push(chunk);
+      length += chunk.length;
+
+      // Once we've read enough bytes, terminate the stream
```

Comment:
I'd prefer not to use another dep for this, especially since it's a rather simple operation to do. Also `slice-stream` doesn't end / close / destroy the stream once the required slice is read.